sklearn_numba_dpex/kmeans/kernels/utils.py

```diff
@@ -25,6 +25,150 @@
+        item_idx = dpex.get_local_id(0)
+        relocated_idx = group_idx // n_work_groups_for_cluster
+        feature_idx = ((group_idx % n_work_groups_for_cluster) * work_group_size) + item_idx
+
+        if feature_idx >= n_features:
+            return
+
+        relocated_cluster_idx = empty_clusters_list[relocated_idx]
+        new_location_X_idx = samples_far_from_center[n_selected_gt_threshold - relocated_idx]
+        new_location_previous_assignment = assignments_idx[new_location_X_idx]
+
+        new_centroid_value = X_t[feature_idx, new_location_X_idx]
+        new_location_weight = sample_weight[new_location_X_idx]
+        X_centroid_addend = new_centroid_value * new_location_weight
+
+        # Cancel the contribution to the updated centroids of the sample that was once
+        # assigned to new_location_previous_assignment but is now assigned to the
+        # cluster of the centroids that relocates to this sample
+        dpex.atomic.sub(
+            centroids_t,
+            (feature_idx, new_location_previous_assignment),
+            X_centroid_addend
+            )
+
+        # The relocated centroid has only one contribution now, which is the sample
+        # to which it has been relocated to
+        centroids_t[feature_idx, relocated_cluster_idx] = X_centroid_addend
+
+        # Likewise, we update the weights in the clusters
+        if feature_idx == zero:
+            per_sample_inertia[new_location_X_idx] = zero
+            dpex.atomic.sub(
+                cluster_sizes,
+                new_location_previous_assignment,
+                new_location_weight
+            )
+            cluster_sizes[relocated_cluster_idx] = new_location_weight
+
+    return relocate_empty_clusters[global_size, work_group_size]
+
+
+@lru_cache
+def make_select_samples_far_from_centroid_kernel(
+    threshold, n_selected, n_samples, work_group_size
+):
+    global_size = math.ceil(n_selected / work_group_size) * work_group_size
+    zero = np.int64(0)
+    one = np.int32(1)
+    max_n_selected_gt_threshold = np.int32(n_selected - 1)
+    min_n_selected_eq_threshold = np.int32(2)
+    max_n_selected_eq_threshold = np.int32(n_selected + 1)
+
+    @dpex.kernel
+    # fmt: off
+    def select_samples_far_from_centroid(
+            distance_to_centroid,           # IN       (n_samples,)
+            selected_samples_idx,           # OUT      (n_samples,)
+            n_selected_gt_threshold,        # OUT      (1, )
+            n_selected_eq_threshold,        # OUT      (1, )
+        ):
+    # fmt: on
+        """This kernel writes in selected_samples_idx the idx of the top n_selected
+        items in distance_to_centroid with highest values.
+
+        threshold is expected to have been pre-computed (by partitioning) such that
+        there are at most (n_selected-1) values that are strictly greater than
+        threshold, and at least n_selected values that are greater or equal than
+        threshold.
```

Comment:
I find numerical value being parenthesised unusual.

Would it be better to have them backquoted? E.g.
```suggestion
        threshold is expected to have been pre-computed (by partitioning) such that
        there are at most `n_selected-1` values that are strictly greater than
        threshold, and at least n_selected values that are greater or equal than
        threshold.
```

This comment applies to others docstrings.

Reply:
:heavy_check_mark: I apply this but about the others IDK how to detect all parenthesis expression in docstrings :-/

Reply:
Ah, I meant parenthesised expressions in docstrings and comments introduced here can be manually be changed to use backquotes. Once again that's a nit.

Reply:
caught a few more